src/classes/RD2_ValidationService.cls

```diff
@@ -0,0 +1,114 @@
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Recurring Donations
+* @description Recurring Donations Validation Service
+*
+*/
+public with sharing class RD2_ValidationService {
+
+    /***
+    * @description Validates missing data on Recurring Donation
+    * @param rds the Recurring Donation to validate
+    */
+    public void validate(List<npe03__Recurring_Donation__c> rds) {
+
+        Set<Id> accountIds = new Set<Id>();
+        Set<Id> contactIds = new Set<Id>();
+
+        for (npe03__Recurring_Donation__c rd : rds) {
+            if (rd.npe03__Organization__c != null) {
+                accountIds.add(rd.npe03__Organization__c);
+                contactIds.add(rd.npe03__Contact__c);
+            }
+        }
+        accountIds.remove(null);
+        contactIds.remove(null);
+
+        Map<id,Account> accountsById = new Map<Id,Account>([
+            SELECT Id, npe01__SYSTEMIsIndividual__c,
+            (
+                SELECT Id
+                FROM Contacts
+                WHERE Id IN :contactIds
+            )
+            FROM Account
+            WHERE Id IN :accountIds
+            ]);
+
+        for (npe03__Recurring_Donation__c rd : rds) {
+            if (rd.npe03__Organization__c == null && rd.npe03__Contact__c == null) {
+                rd.addError(System.Label.RD_DonorIsRequired);
+                return;
+            }
+            validate(rd, accountsById.get(rd.npe03__Organization__c));
+        }
+    }
+
+    /***
+    * @description Validates missing account or contact data on Recurring Donation
+    * @param rd the Recurring Donation to validate
+    * @param acct is the associated account for the Recurring Donation
+    */
+    public void validate(npe03__Recurring_Donation__c rd, Account acct) {
+
+        if (acct == null) {
+            return;
+        }
+
+        if (acct.npe01__SYSTEMIsIndividual__c) {
+            if (rd.npe03__Contact__c == null) {
+                rd.addError(System.Label.RD_ContactMustBelongToAccount);
+            } else if (!validateContact(rd, acct.Contacts) || acct.Contacts.size() == 0) {
+                rd.addError(System.Label.RD_ContactMustBelongToAccount);
+            }
+        }
+    }
+
+    /***
+    * @description Validates that the Contact on the RD is a member of the Household Account.
+    * @param rd the Recurring Donation to validate
+    * @param Contacts are the contacts related to the account.
+    */
+    @TestVisible
+    private Boolean validateContact(npe03__Recurring_Donation__c rd, List<Contact> Contacts) {
```

Comment:
To improve performance (the above is a `for` loop inside the `for` loop), this method can be replaced by the following:
```
    /***
    * @description Extracts record Ids
    * @param records 
    * @return Set<Id>
    */
    public Set<Id> extractIds(List<SObject> records) {
        if (records == null || records.isEmpty()) {
            return new Set<Id>();
        }
        return new Map<Id, SObject>(records).keySet();
    }
```

so the `validate()` method would verify Contact on the Household as:
```
    if (rd.npe03__Contact__c == null) {
        rd.addError(System.Label.RD_ContactMustBelongToAccount);
        
    } else if (!extractIds(acc.Contacts).contains(rd.npe03__Contact__c)) {
        rd.addError(System.Label.RD_ContactMustBelongToAccount);
    }
```

Reply:
@RadSFDeveloper , I like this idea!